### ISSUE TITLE
log-scale Breaks Are Not Always Useful

### ISSUE THREAD

**Issue by billdenney** [r-lib/scales#100]
This is tidyverse/ggplot2#2269 and somewhat related to #9.

When plotting with ggplot2 and using log-scale axes, breaks don't always show up.  This is related to the fact that log_breaks doesn't always provide breaks within the range.  A specific example would be:

```r
> scales::log_breaks()(2:3)
[1]  1 10
```

When a plot is generated, both 1 and 10 are often outside of the range, and the number of breaks requested is 5 when only 2 are returned which also appears to be a bug.  More useful (in most scenarios that I think of) would be to select human-readable numbers that are approximately within the range of interest.

Specifically, I think that the preferred scales (at least for log 10) would minimize the number of decimal significant figures while matching the requested number of breaks with at least `n - 2` of the breaks within the range of interest. (`n - 2` rather than `n` to allow for the fact that being just outside the range of interest could still show up on the scale.)  The right metric on scales other than log10 may be to minimize the number of significant figures in the exponent.

For example, the following still doesn't match the requested number of breaks being 5 nor are the breaks within the range.

```r
> scales::log_breaks(base=exp(1))(c(2, 2.0001))
[1] 1.000000 2.718282
```

This function appears to better align with the hope of showing log-scale breaks even when the range does not align with the desire of generating breaks for plotting:  The number of requested breaks are returned, and the breaks are (approximately) equally spaced on the log scale:

```r
log_breaks <- function(n=5, base=10) {
  function(x) {
    rng <- log(range(x, na.rm = TRUE), base = base)
    min <- floor(rng[1])
    max <- ceiling(rng[2])
    
    if (max == min) {
      ret <- base ^ min
    } else if ((max - min) <= (n - 2)) {
      # Approximately minimize the number of log10 significant figures
      ret <- seq(rng[1], rng[2], length.out=n)
      rounding_digits <- -floor(log10(diff(base^ret)))
      rounding_digits <- c(rounding_digits[1], rounding_digits)
      ret <- round(base^ret, rounding_digits)
    } else {
      by <- floor((max - min) / n) + 1
      ret <- base ^ seq(min, max, by = by)
    }
    ret
  }
}
```

This function yields the following results (compared to the current version in the scales package).

```r
> log_breaks(base=exp(1))(c(2.0001, 2.000111111))
[1] 2.000100 2.000103 2.000106 2.000108 2.000111
> scales::log_breaks(base=exp(1))(c(2.0001, 2.000111111))
[1] 1.000000 2.718282

> log_breaks(base=10)(c(2.0001, 2.000111111))
[1] 2.000100 2.000103 2.000106 2.000108 2.000111
> scales::log_breaks(base=10)(c(2.0001, 2.000111111))
[1]  1 10

> log_breaks(base=2)(c(2.0001, 2.000111111))
[1] 2.000100 2.000103 2.000106 2.000108 2.000111
> scales::log_breaks(base=2)(c(2.0001, 2.000111111))
[1] 2 4

> scales::log_breaks(base=2)(c(2.0001, 100))
[1]   2   8  32 128
> log_breaks(base=2)(c(2.0001, 100))
[1]   2   8  32 128

# Maybe worse here, could be tweaked by modifying the else if statement
> log_breaks(base=2)(c(2.0001, 10))
[1]  2  3  4  7 10
> scales::log_breaks(base=2)(c(2.0001, 10))
[1]  2  4  8 16

# Also note that with the proposed function results are effectively
# log-scale independent in these smaller ranges
> log_breaks(base=10)(c(2.0001, 14))
[1]  2  3  5  9 14
> log_breaks(base=2)(c(2.0001, 14))
[1]  2  3  5  9 14
> log_breaks(base=1e9)(c(2.0001, 14))
[1]  2  3  5  9 14
```

**Comment by has2k1** [comment]
How about falling back to `extended_breaks`, I think it would make sense to do so for base 10.

**Comment by billdenney** [comment]
`extended_breaks` would work for base 10 when the values are approximately within about one order of magnitude (and, that is probably worth special case).  But, I don't think it would work well when for example n=10 and the range is 3 orders of magnitude because the returned breaks would be clustered toward the upper end of the log-range.

**Comment by has2k1** [comment]
Yes `extended_breaks` only for that special case.

**Comment by billdenney** [comment]
So I did a quick try on using extended_breaks with an update to the version below.  I had to eliminate zero from the suggested values of `extended_breaks`.  With 1 order of magnitude in the range, the ticks at the bottom of the range are about 3-fold farther apart than the top of the range (at 1.5 orders of magnitude, they are about 4-fold farther apart), but also by dropping zero.

My preference is initially suggested version, but I think that either will suffice.

Should I generate a pull request with this function (or the previous one)?

```r
log_breaks <- function(n=5, base=10) {
  function(x) {
    rng <- log(range(x, na.rm = TRUE), base = base)
    min <- floor(rng[1])
    max <- ceiling(rng[2])
    
    if (max == min) {
      ret <- base ^ min
    } else if ((base == 10) &
               ((rng[2] - rng[1]) < 1)) {
      # Special case if log10 is used, and the range is reasonably
      # small, fall back to extended_breaks.
      ret <- setdiff(extended_breaks(n=n)(x), 0)
    } else if ((max - min) <= (n - 2)) {
      # Approximately minimize the number of log10 significant figures
      ret <- seq(rng[1], rng[2], length.out=n)
      rounding_digits <- -floor(log10(diff(base^ret)))
      rounding_digits <- c(rounding_digits[1], rounding_digits)
      ret <- round(base^ret, rounding_digits)
    } else {
      by <- floor((max - min) / n) + 1
      ret <- base ^ seq(min, max, by = by)
    }
    ret
  }
}
```

**Comment by hadley** [comment]
Another thought:

``` r
library(ggplot2)

ggplot(diamonds, aes(carat, price)) +
  geom_hex() + 
  scale_x_log10(limits = c(0.2, 5), breaks = c(0.1, 0.2, 0.5, 1, 2, 5)) + 
  scale_y_log10(limits = c(300, 19000))
#> Warning: Removed 1 rows containing non-finite values (stat_binhex).
#> Warning: Removed 14 rows containing missing values (geom_hex).
```

![](https://i.imgur.com/1XiViOs.png)

Created on 2018-05-10 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by billdenney** [comment]
@hadley , I often use a solution similar to that (`breaks=outer(c(1, 2, 5), 10^(-5:5), FUN="*")` is what I typically do).  My hope is that we will end up with something that works across all data, even if the range is 0.002 to 0.0021.

**Comment by dpseidel** [comment]
Addressed and closed by #117
